game: make game status modular

lahm86 opened the pull request:
Resolves #1070.
Resolves #1074.

#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This removes the global game status and instead exposes functions for setting, adding, removing and reverting status. Callers still need to manage the call order, but it removes the various bitwise operations. Setting is done for main changes, like entering the inventory; adding/removing is intended for "sub" changes within that status (like `GMS_IN_INVENTORY_HEALTH`); restoring reverts back to a cached value created on the last `Game_SetStatus` call - so is intended for the likes of exiting the inventory.

The two linked issues themselves were caused by the inventory not reverting to `GMS_IN_GAME`. In `Random_GetDraw`, the check was looking for this value precisely (to indicate that the player was not in the inventory, pause or stats screen and hence to freeze effects).

lahm86 committed:
game: make game status modular

This removes the global game status and instead exposes functions for
setting, adding, removing and reverting status. Callers still need to
manage the call order, but it removes the various bitwise operations.
This resolves an issue where opening the inventory would not reverting
the status to GMS_IN_GAME to 0, which the check in Random_GetDraw looks
for. As a result, RNG was not progressing for such things as FX and
SFX.

Resolves #1070
Resolves #1074

rr- commented:
Outsourcing the state management to a special family of functions is a good call. We still treat the status sometimes as a bit field, and sometimes as an enum though. Perhaps we could change `GAME_STATE` to be a simple enum (main menu, game, inventory, pause, stats) and move flags such as `GMS_MAIN_MENU_SHOWN` to a new entity? I think this would simplify mental model around the game state. This new entity could also hold ng+ flag and possibly more – I think it would make sense to group as much together as possible.

rr- commented:
Something like this:
```patch
diff --git a/src/game/game.h b/src/game/game.h
index a792d14f..28aa4092 100644
--- a/src/game/game.h
+++ b/src/game/game.h
@@ -5,11 +5,8 @@
 #include <stdbool.h>
 #include <stdint.h>
 
-bool Game_HasStatus(GAME_STATUS status);
-void Game_SetStatus(GAME_STATUS status);
-void Game_AddStatus(GAME_STATUS status);
-void Game_RemoveStatus(GAME_STATUS status);
-void Game_RestoreStatus(void);
+GAME_STATUS Game_GetStatus(void);
+void Game_SetStatus(GAME_STATUS stauts);
 
 bool Game_Start(int32_t level_num, GAMEFLOW_LEVEL_TYPE level_type);
 int32_t Game_Stop(void);
diff --git a/src/game/game/game.c b/src/game/game/game.c
index 288de75d..1ffc8c72 100644
--- a/src/game/game/game.c
+++ b/src/game/game/game.c
@@ -31,8 +31,7 @@
 
 static const int32_t m_AnimationRate = 0x8000;
 static int32_t m_FrameCount = 0;
-static GAME_STATUS m_CurrentStatus = GMS_IN_GAME;
-static GAME_STATUS m_PreviousStatus = GMS_IN_GAME;
+static GAME_STATUS m_CurrentStatus = GS_INITIAL;
 
 static int32_t Game_Control(int32_t nframes, GAMEFLOW_LEVEL_TYPE level_type);
 
@@ -155,7 +154,7 @@ void Game_ProcessInput(void)
         Lara_UseItem(O_BIGMEDI_OPTION);
     }
 
-    if (g_Config.enable_buffering && Game_HasStatus(GMS_IN_GAME)) {
+    if (g_Config.enable_buffering && Game_GetStatus() == GS_IN_GAME) {
         if (g_Input.toggle_bilinear_filter) {
             FRAME_BUFFER(toggle_bilinear_filter);
         } else if (g_Input.toggle_perspective_filter) {
@@ -166,36 +165,20 @@ void Game_ProcessInput(void)
     }
 }
 
-bool Game_HasStatus(GAME_STATUS status)
+GAME_STATUS Game_GetStatus(void)
 {
-    return m_CurrentStatus & status;
+    return m_CurrentStatus;
 }
 
 void Game_SetStatus(GAME_STATUS status)
 {
-    m_PreviousStatus = m_CurrentStatus;
     m_CurrentStatus = status;
 }
 
-void Game_AddStatus(GAME_STATUS status)
-{
-    m_CurrentStatus |= status;
-}
-
-void Game_RemoveStatus(GAME_STATUS status)
-{
-    m_CurrentStatus &= ~status;
-}
-
-void Game_RestoreStatus(void)
-{
-    m_CurrentStatus = m_PreviousStatus;
-}
-
 bool Game_Start(int32_t level_num, GAMEFLOW_LEVEL_TYPE level_type)
 {
     g_GameInfo.current_level_type = level_type;
-    Game_SetStatus(GMS_IN_GAME);
+    Game_SetStatus(GS_IN_GAME);
 
     switch (level_type) {
     case GFL_SAVED:
diff --git a/src/game/game/game_pause.c b/src/game/game/game_pause.c
index 50cf6587..84ff578c 100644
--- a/src/game/game/game_pause.c
+++ b/src/game/game/game_pause.c
@@ -152,7 +152,8 @@ bool Game_Pause(void)
 
     Music_Pause();
     Sound_PauseAll();
-    Game_SetStatus(GMS_IN_PAUSE);
+    const GAME_STATUS old_status = Game_GetStatus();
+    Game_SetStatus(GS_IN_PAUSE);
 
     Output_FadeToSemiBlack(true);
     int32_t select = Game_Pause_Loop();
@@ -162,6 +163,6 @@ bool Game_Pause(void)
     Sound_UnpauseAll();
     Requester_Remove(&m_PauseRequester);
     Game_Pause_RemoveText();
-    Game_RestoreStatus();
+    Game_SetStatus(old_status);
     return select < 0;
 }
diff --git a/src/game/gameflow.c b/src/game/gameflow.c
index a66c0513..51763ff3 100644
--- a/src/game/gameflow.c
+++ b/src/game/gameflow.c
@@ -1301,8 +1301,7 @@ GameFlow_InterpretSequence(int32_t level_num, GAMEFLOW_LEVEL_TYPE level_type)
                 break;
             }
 
-            if (level_type == GFL_TITLE && !Game_HasStatus(GMS_GAME_MENU_SHOWN)
-                && !g_Config.enable_eidos_logo) {
+            if (Game_GetStatus() == GS_INITIAL && !g_Config.enable_eidos_logo) {
                 break;
             }
 
diff --git a/src/game/inventory/inventory.c b/src/game/inventory/inventory.c
index 6c0d94ac..f4ab0d71 100644
--- a/src/game/inventory/inventory.c
+++ b/src/game/inventory/inventory.c
@@ -1011,10 +1011,12 @@ static int32_t Inv_ConstructAndDisplay(int inv_mode)
 
 int32_t Inv_Display(int inv_mode)
 {
-    Game_AddStatus(GMS_GAME_MENU_SHOWN);
-    Game_SetStatus(GMS_IN_INVENTORY);
+    const GAME_STATUS old_status = Game_GetStatus();
+    Game_SetStatus(
+        g_CurrentLevel == g_GameFlow.title_level_num ? GS_IN_MAIN_MENU
+                                                     : GS_IN_INVENTORY);
     int32_t inv_result = Inv_ConstructAndDisplay(inv_mode);
-    Game_RestoreStatus();
+    Game_SetStatus(old_status);
 
     if (g_Config.enable_buffering) {
         g_OldInputDB.any = 0;
diff --git a/src/game/inventory/inventory_ring.c b/src/game/inventory/inventory_ring.c
index 6cca93f5..1c0a2e28 100644
--- a/src/game/inventory/inventory_ring.c
+++ b/src/game/inventory/inventory_ring.c
@@ -1,7 +1,6 @@
 #include "game/inventory/inventory_ring.h"
 
 #include "config.h"
-#include "game/game.h"
 #include "game/gameflow.h"
 #include "game/inventory.h"
 #include "game/inventory/inventory_vars.h"
@@ -299,13 +298,13 @@ void Inv_Ring_Active(INVENTORY_ITEM *inv_item)
         } else if (g_Config.healthbar_location == BL_BOTTOM_RIGHT) {
             Text_Hide(m_InvDownArrow2, true);
         }
-        Game_AddStatus(GMS_IN_INVENTORY_HEALTH);
+        g_GameInfo.inv_showing_medpack = true;
     } else {
         Text_Hide(m_InvUpArrow1, false);
         Text_Hide(m_InvUpArrow2, false);
         Text_Hide(m_InvDownArrow1, false);
         Text_Hide(m_InvDownArrow2, false);
-        Game_RemoveStatus(GMS_IN_INVENTORY_HEALTH);
+        g_GameInfo.inv_showing_medpack = false;
     }
 }
 
diff --git a/src/game/objects/traps/lightning_emitter.c b/src/game/objects/traps/lightning_emitter.c
index d1d68a14..5c871804 100644
--- a/src/game/objects/traps/lightning_emitter.c
+++ b/src/game/objects/traps/lightning_emitter.c
@@ -214,7 +214,7 @@ void LightningEmitter_Draw(ITEM_INFO *item)
 
     for (int i = 0; i < LIGHTNING_STEPS; i++) {
         PHD_VECTOR *pos = &l->wibble[i];
-        if (Game_HasStatus(GMS_IN_GAME)) {
+        if (Game_GetStatus() == GS_IN_GAME) {
             pos->x += (Random_GetDraw() - PHD_90) * LIGHTNING_RND;
             pos->y += (Random_GetDraw() - PHD_90) * LIGHTNING_RND;
             pos->z += (Random_GetDraw() - PHD_90) * LIGHTNING_RND;
@@ -268,7 +268,7 @@ void LightningEmitter_Draw(ITEM_INFO *item)
 
         for (j = 0; j < steps; j++) {
             PHD_VECTOR *pos = l->shoot[i];
-            if (Game_HasStatus(GMS_IN_GAME)) {
+            if (Game_GetStatus() == GS_IN_GAME) {
                 pos->x += (Random_GetDraw() - PHD_90) * LIGHTNING_RND;
                 pos->y += (Random_GetDraw() - PHD_90) * LIGHTNING_RND;
                 pos->z += (Random_GetDraw() - PHD_90) * LIGHTNING_RND;
diff --git a/src/game/overlay.c b/src/game/overlay.c
index 223aa6b9..c63a2f40 100644
--- a/src/game/overlay.c
+++ b/src/game/overlay.c
@@ -229,7 +229,7 @@ static void Overlay_BarGetLocation(
             - m_BarOffsetY[bar_info->location];
     }
 
-    if (Game_HasStatus(GMS_IN_INVENTORY)
+    if (Game_GetStatus() == GS_IN_INVENTORY
         && (bar_info->location == BL_TOP_CENTER
             || bar_info->location == BL_BOTTOM_CENTER)) {
         double scale_bar_to_text =
@@ -727,9 +727,10 @@ void Overlay_DrawFPSInfo(void)
             elapsed = Clock_GetMS();
         }
 
-        bool inv_health_showable = Game_HasStatus(GMS_IN_INVENTORY_HEALTH)
+        bool inv_health_showable = Game_GetStatus() == GS_IN_INVENTORY
+            && g_GameInfo.inv_showing_medpack
             && m_HealthBar.location == BL_TOP_LEFT;
-        bool game_bar_showable = !Game_HasStatus(GMS_IN_INVENTORY)
+        bool game_bar_showable = Game_GetStatus() == GS_IN_GAME
             && (m_HealthBar.location == BL_TOP_LEFT
                 || m_AirBar.location == BL_TOP_LEFT
                 || m_EnemyBar.location == BL_TOP_LEFT);
@@ -739,7 +740,7 @@ void Overlay_DrawFPSInfo(void)
             y = text_height
                 + scale_fps_to_bar * (y + m_BarOffsetY[BL_TOP_LEFT]);
         } else if (
-            Game_HasStatus(GMS_IN_INVENTORY) && g_GameInfo.inv_ring_above) {
+            Game_GetStatus() == GS_IN_INVENTORY && g_GameInfo.inv_ring_above) {
             y += (text_height * 2) + text_inv_offset_y;
         } else {
             y += text_height;
diff --git a/src/game/random.c b/src/game/random.c
index f94acea5..3bd4b53e 100644
--- a/src/game/random.c
+++ b/src/game/random.c
@@ -27,7 +27,7 @@ void Random_SeedDraw(int32_t seed)
 
 int32_t Random_GetDraw(void)
 {
-    if (Game_HasStatus(GMS_IN_GAME)) {
+    if (Game_GetStatus() == GS_IN_GAME) {
         m_RandDraw = 0x41C64E6D * m_RandDraw + 0x3039;
     }
     return (m_RandDraw >> 10) & 0x7FFF;
diff --git a/src/game/stats.c b/src/game/stats.c
index f01c29fe..b3f1f167 100644
--- a/src/game/stats.c
+++ b/src/game/stats.c
@@ -307,7 +307,8 @@ void Stats_Show(int32_t level_num)
         return;
     }
 
-    Game_SetStatus(GMS_IN_STATS);
+    const GAME_STATUS old_status = Game_GetStatus();
+    Game_SetStatus(GS_IN_STATS);
 
     char buf[100];
     char time_str[100];
@@ -424,7 +425,7 @@ void Stats_Show(int32_t level_num)
     }
 
     Output_FadeReset();
-    Game_RestoreStatus();
+    Game_SetStatus(old_status);
 
     for (int i = 0; i < MAX_TEXTSTRINGS; i++) {
         cur_txt = &all_txt[i];
diff --git a/src/global/types.h b/src/global/types.h
index a8ca7c47..33c68bcd 100644
--- a/src/global/types.h
+++ b/src/global/types.h
@@ -1566,12 +1566,12 @@ typedef struct RESUME_INFO {
 } RESUME_INFO;
 
 typedef enum GAME_STATUS {
-    GMS_IN_GAME = 1 << 0,
-    GMS_IN_INVENTORY = 1 << 1,
-    GMS_IN_PAUSE = 1 << 2,
-    GMS_IN_STATS = 1 << 3,
-    GMS_IN_INVENTORY_HEALTH = 1 << 4,
-    GMS_GAME_MENU_SHOWN = 1 << 5,
+    GS_INITIAL = 0,
+    GS_IN_MAIN_MENU = 1,
+    GS_IN_GAME = 2,
+    GS_IN_INVENTORY = 3,
+    GS_IN_PAUSE = 4,
+    GS_IN_STATS = 5,
 } GAME_STATUS;
 
 typedef struct GAME_INFO {
@@ -1589,6 +1589,7 @@ typedef struct GAME_INFO {
     bool remove_scions;
     bool remove_ammo;
     bool remove_medipacks;
+    bool inv_showing_medpack;
     bool inv_ring_above;
 } GAME_INFO;
 
```

lahm86 committed:
game: simplify status logic

Converted GAME_STATUS to an enum, so bitwise checks have been replaced
accordingly. Separate fields and checks made for previous sub-statuses,
like inv_showing_medpack and GS_INITIAL for checking the Eidos logo
option.

walkawayy reviewed (approved):
All testing seems good to me!
